### PR TITLE
KAFKA-4671: Fix Streams window retention policy

### PR DESCRIPTION
**Pull request by mjsax**

### CODE REVIEW

**Comment by mjsax** [comment]
Extracted code changes from #2337 

@dguy @enothereska @guozhangwang 

Should go into `0.10.2`, too.

**Review by dguy — COMMENTED** [review]
Just a general comment on the test names. I see you change some test names in `JoinWindowsTest` to be more descriptive - thanks.
However, in the other tests, i.e., `SessionWindowsTest`, `TimeWindowsTest` etc, we still have non-descriptive test names. It is great that we have the tests, but it is much better if we can just read the test names and work out what it is doing. This is why i almost always start the name of a test with should. I.e, what should it do? I'm sorry to bang on about this, but i find it super useful and important. If the test names are descriptive then you can usually understand what a class does without having to read the code.

File: streams/src/main/java/org/apache/kafka/streams/kstream/JoinWindows.java [hunk 1]
```diff
@@ -108,16 +104,31 @@ public long size() {
     }
 
     @Override
-    public final boolean equals(Object o) {
+    public Windows<Window> until(final long duration) throws IllegalArgumentException {
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/JoinWindows.java]: Change return type to `JoinWindows` (same for all other classed inheriting from `Windows`)

**Comment by mjsax** [comment]
Updated this.
@dguy @enothereska @guozhangwang

**Comment by mjsax** [comment]
retest this

**Comment by mjsax** [comment]
test this please

**Review by dguy — APPROVED** [review]
Thanks @mjsax LGTM

**Comment by enothereska** [comment]
LGTM

File: streams/src/main/java/org/apache/kafka/streams/kstream/SessionWindows.java [hunk 2]
```diff
@@ -60,12 +60,12 @@
 @InterfaceStability.Unstable
 public class SessionWindows {
 
-    private final long gapMs;
-    private long maintainDurationMs;
+    private final long gap;
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/kstream/SessionWindows.java]: Not clear why removing `Ms` suffix?

File: streams/src/main/java/org/apache/kafka/streams/kstream/TimeWindows.java [hunk 3]
```diff
@@ -93,18 +87,21 @@ public static TimeWindows of(long size) {
      *                 {@link org.apache.kafka.streams.processor.TimestampExtractor}.
      * @return a new window definition
      */
-    public TimeWindows advanceBy(long interval) {
-        return new TimeWindows(this.size, interval);
+    public TimeWindows advanceBy(final long advance) {
+        if (!(0 < advance && advance <= size)) {
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/kstream/TimeWindows.java]: nit: `if (advance <= 0 || advance > size)`

File: streams/src/main/java/org/apache/kafka/streams/kstream/Windows.java [hunk 4]
```diff
@@ -28,15 +27,15 @@
 
     private static final int DEFAULT_NUM_SEGMENTS = 3;
 
-    static final long DEFAULT_MAINTAIN_DURATION = 24 * 60 * 60 * 1000L;   // one day
+    static final long DEFAULT_MAINTAIN_DURATION = 24 * 60 * 60 * 1000L; // one day (if time is represented as ms)
 
-    private long maintainDurationMs;
+    private long maintainDuration;
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/kstream/Windows.java]: Ditto above.

File: streams/src/main/java/org/apache/kafka/streams/kstream/JoinWindows.java [hunk 5]
```diff
@@ -108,16 +104,31 @@ public long size() {
     }
 
     @Override
-    public final boolean equals(Object o) {
+    public JoinWindows until(final long duration) throws IllegalArgumentException {
+        if (duration < size()) {
+            throw new IllegalArgumentException("Window retention time (duration) cannot be smaller than the window " +
+                "size.");
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/kstream/JoinWindows.java]: nit: this can be a single line.

**Comment by mjsax** [comment]
@guozhangwang Updated.

**Comment by mjsax** [comment]
@guozhangwang Updated all variables with `Ms` suffix.

**Comment by guozhangwang** [comment]
retest this please

**Comment by guozhangwang** [comment]
LGTM. Merged to trunk and 0.10.2.

**Pull request closed without merging**